src/SecurityBundle/src/CmsGoogleLogin/GoogleUserRegistrationService.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace ChameleonSystem\SecurityBundle\CmsGoogleLogin;
+
+use ChameleonSystem\CoreBundle\Interfaces\GuidCreationServiceInterface;
+use ChameleonSystem\SecurityBundle\CmsUser\CmsUserDataAccess;
+use ChameleonSystem\SecurityBundle\CmsUser\CmsUserModel;
+use League\OAuth2\Client\Provider\GoogleUser;
+
+class GoogleUserRegistrationService implements GoogleUserRegistrationServiceInterface
+{
+    public function __construct(
+        private readonly CmsUserDataAccess $cmsUserDataAccess,
+        private readonly GuidCreationServiceInterface $guidService,
+        private readonly array $allowedDomains = []
+    ) {
+    }
+
+    public function register(GoogleUser $googleUser): CmsUserModel
+    {
+        if ($this->exists($googleUser)) {
+            throw new \Exception('User already exists');
+        }
+
+        $hostedDomain = $googleUser->getHostedDomain();
+        $email = $googleUser->getEmail();
+        $mailDomain = mb_substr($email, mb_strpos($email, '@')+1);
+        $baseUserName = $this->allowedDomains[$hostedDomain] ?? $this->allowedDomains[$mailDomain] ?? null;
+        if (null === $baseUserName) {
+            throw new \Exception('You may only register google users from the following domains: '.implode(', ', array_keys($this->allowedDomains)));
+        }
+        $user = $this->cmsUserDataAccess->loadUserByIdentifier($baseUserName)
+                ->withId($this->guidService->findUnusedId('cms_user'))
+                ->withUserIdentifier($googleUser->getEmail())
+                ->withDateModified(new \DateTimeImmutable())
+                ->withCompany($googleUser->getHostedDomain())
```

Comment:
It is in a way. The hosted domain is only for google app company users - it may not be the company itself, but the closest thing we have in the google user